plugin/rewrite: streamline the ResponseRule handling.

mandelsoft opened the pull request:
<!--
Thank you for contributing to CoreDNS!
Please provide the following information to help us make the most of your pull request:
-->

### 1. Why is this pull request needed and what does it do?

The functionality of a response rule is now completely encapsulated behind
a `ResponseRule` interface. This significantly simplifies the complete
processing flow, it enables more flexible response handling and it is possible
to eliminate lots of state flags, ifs and switches.

Based on the new flexibility the pull request also enables to support a
response name rewrite for all name rewrite types.
To be compatible, an explicit `answer auto` option is added for this.

Using this feature it is also possible now to rewrite a complete sub domain
hierarchy to a single domain name combined with a correct answer rewrite (coredns#2389).

The tests have been improved to test the new cases but also lots of older untested cases.

### 2. Which issues (if any) are related?
#2389, #2202

### 3. Which documentation changes (if any) need to be made?

The documentation of the rewrite plugin has been adapted to reflect the new answer option.

### 4. Does this introduce a backward incompatible change or deprecation?

No, the former behaviour completely remains. It just adds a new answer option for the rewrite plugin that can be used to influence the behaviour of all name rewrite types.

chrisohaver commented on plugin/rewrite/README.md:
```diff
@@ -127,6 +128,28 @@ ftp.service.us-west-1.consul. 0    IN A    10.30.30.30
+Any occurrence of the rewritten question in the answer is mapped
+back to the original value before the rewrite.
+
+Please not that answers for rewrites of type `exact` are always rewritten.
```
```suggestion
Please note that answers for rewrites of type `exact` are always rewritten.
```

miekg commented on plugin/rewrite/name.go:
```diff
@@ -58,62 +114,62 @@ const (
 	}
 	state.Req.Question[0].Name = s
-	return RewriteDone
+	return rule.Response(state)
```
why does this need to be returned? `rule.Response(state)` can also be called outside of this method

mandelsoft replied:
This is a dedicated implementation for the name rewrite types (the other types (can) do it differently), that need to dynamically handle the auto and non-auto case. I've separated this into this response function, because it is identical for all name types. `Response` is just an attribute, here. The interface function to be used from the outside is still the `Rewrite` method of a rule object, which returns a dedicated, possibly request specific response rule . Because of this feature to provide a request specific response rule, a `GetResponse()` method cannot be used for the rewrite rule type object. The rewrite object must have the full control depending on the rewrite request to provide a dedicated response rule.
Basically all the attributes should be internal(lower case), but they were public so so far, therefore I didn't changed this.

mandelsoft replied:
May be it is better to encapsulate the common behaviour of those name rules into a dedicated base struct used for the name rule types. I could provide this, It could also handle the `Mode()` function.

mandelsoft replied:
I've changed this now and introduced a common base part, which allowed to get rid of even more coding

miekg replied:
making members/functions private is OK w/ me. I think this was mostly a side effect of style. There should be nothing consuming this as an API

miekg commented on plugin/rewrite/name.go:
```diff
@@ -140,45 +193,56 @@ func newNameRule(nextAction string, args ...string) (Rule, error) {
 	}
 
 	if len(args) < 7 {
+		_ = auto
```
why?

mandelsoft replied:
sorry , an artefact from testing, I'll remove it

miekg commented on plugin/rewrite/name.go:
```diff
@@ -140,45 +193,56 @@ func newNameRule(nextAction string, args ...string) (Rule, error) {
 		case ExactMatch:
-			rewriteAnswerFromPattern, err := isValidRegexPattern(rewriteQuestionTo, rewriteQuestionFrom)
+			_, err := isValidRegexPattern(rewriteQuestionTo, rewriteQuestionFrom)
 			if err != nil {
```
this if can no be folded into the check, saving a line

miekg commented on plugin/rewrite/name_test.go:
```diff
@@ -75,13 +75,244 @@ func TestRewriteNamePrefixSuffix(t *testing.T) {
 		expectedFail bool
 	}{
-		{"stop", []string{"exact", "srv3.coredns.rocks", "srv4.coredns.rocks"}, false},
+		{"stop", []string{"exact", "rv3.coredns.rocks", "srv4.coredns.rocks"}, false},
```
why this change?

mandelsoft replied:
not intended, sorry, I'll revert to the original one

miekg commented on plugin/rewrite/reverter.go:
```diff
@@ -23,15 +25,16 @@ type ResponseRule struct {
 
 // NewResponseReverter returns a pointer to a new ResponseReverter.
-func NewResponseReverter(w dns.ResponseWriter, r *dns.Msg) *ResponseReverter {
+func NewResponseReverter(w dns.ResponseWriter, r *dns.Msg, noSaveQuestion bool) *ResponseReverter {
```
```suggestion
func NewResponseReverter(w dns.ResponseWriter, r *dns.Msg, restoreQuestion bool) *ResponseReverter {
```

miekg commented on plugin/rewrite/reverter.go:
```diff
@@ -23,15 +25,16 @@ type ResponseRule struct {
 	return &ResponseReverter{
 		ResponseWriter:   w,
 		originalQuestion: r.Question[0],
+		noSave:           noSaveQuestion,
```
```suggestion
		restore:           restoreQuestion,
```

mandelsoft replied:
Better world be `noRestore`, the default (false) should be the standard behavior, which is restoring the original question in the response handling. For testing purposes this will then be set to true.

miekg commented on plugin/rewrite/reverter.go:
```diff
@@ -40,42 +43,13 @@ func (r *ResponseReverter) WriteMsg(res1 *dns.Msg) error {
 
-	res.Question[0] = r.originalQuestion
-	if r.ResponseRewrite {
+	if !r.noSave {
```
```suggestion
	if !r.restore {
```

miekg commented on plugin/rewrite/rewrite.go:
```diff
@@ -33,34 +33,29 @@ const (
 	Next     plugin.Handler
 	Rules    []Rule
-	noRevert bool
+	noRevert bool // for test
```
oh members that _are_ only used for testing, that sounds iffy. Can this be done in another why. I would very much not have this done this way

mandelsoft replied:
The `noRevert` was already there, it has not been introduced by this PR. I just added another such field, and added the comment telling the reason for this field. I just found no other usage for the already existing field, besides various test cases.

miekg commented on plugin/rewrite/ttl.go:
```diff
@@ -9,81 +9,90 @@ import (
 	NextAction string
 	From       string
-	ResponseRule
+	response   *ttlResponseRule
```
why do these need to be pointers now? I think added a NoopRule can also remove the need to make `nil` the thing that conveys 'noop'. Would also make this PR much smaller, 'cause these changes can all be reverted

mandelsoft replied:
I'll use the same pattern with a `ttlRuleBase` as for the name rules, which simplifies the complete rule implementations.

miekg committed:
plugin/rewrite: streamline the ResponseRule handling.

The functionality of a response rule is now completely encapsulated behind
a `ResponseRule` interface. This significantly simplifies the complete
processing flow, it enables more flexible response handling and it is possible
to eliminate lots of state flags, ifs and switches.

Based on the new flexibility the pull request also enables to support a
response name rewrite for all name rewrite types.
To be compatible, an explicit `answer auto` option is added to support
a best effort response rewrite (name and value).
Additionally now all name rewrite rules support additional name and value
reponse rewrite options.

Using this feature it is also possible now to rewrite a complete sub domain
hierarchy to a single domain name combined with a correct rewrite (#2389).

Signed-off-by: Uwe Krueger <uwe.krueger@sap.com>

miekg commented on .github/workflows/go.test.yml:
```diff
@@ -79,4 +79,4 @@ jobs:
 
     - name: Test Makefile
-      run: make -f Makefile.release build
+      run:  make DOCKER=bla GITHUB_ACCESS_TOKEN=bla -n build docker github-push docker-push -f Makefile.release
```
is this a rebase gone wrong?

mandelsoft replied:
I haven't touched the Makefile, but there is an additional change in the meantime. I just did a new rebase.

miekg commented on man/coredns-rewrite.7:
```diff
@@ -18,7 +18,7 @@ A simplified/easy-to-digest syntax for \fIrewrite\fP is...
 
 .nf
-rewrite [continue|stop] FIELD [FROM TO|FROM TTL]
+rewrite [continue|stop] FIELD [TYPE] [(FROM TO)|TTL] [OPTIONS]
```
revert any changes to this file, will be done automatically from a github workflow

miekg commented on man/coredns-sign.7:
```diff
@@ -8,9 +8,9 @@
 .SH "DESCRIPTION"
```
please revert all these changes

miekg commented on plugin/rewrite/name.go:
```diff
@@ -9,40 +9,117 @@ import (
-	From       string
-	To         string
-	ResponseRule
+////////////////////////////////////////////////////////////////////////////////
```
please remove these `////` blocks, it clutters the godoc

miekg commented on plugin/rewrite/rewrite.go:
```diff
@@ -31,37 +31,29 @@ const (
-	noRevert bool
+	Next      plugin.Handler
+	Rules     []Rule
+	noRevert  bool // for test
```
could you add a TODO here, that we want to get rid of these 'used for testing members' in this struct.

mandelsoft replied:
I could propose a general `RevertPolicy` instead, that can officially be set using the API.
Please give me a trigger to push that proposal. So far, I just added the TODO.

mandelsoft committed:
revert policy

Signed-off-by: Uwe Krueger <uwe.krueger@sap.com>
